File: ruby-gem/lib/models/price.rb
# frozen_string_literal: true

# Class from DFC Business Ontology: #Price
require_relative '../core/semantic_object'
require_relative 'quantitative_value'


module DfcLinkmlConnector
  module Models
    class Price < QuantitativeValue
      SEMANTIC_TYPE = "dfc-b:Price".freeze

      # @return [Float]
      attr_accessor :vat_rate

      # @return [String]
      attr_accessor :is_price_of

      # @param semanticId [String]
      # @param date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, value: nil, unit: nil, vatRate: nil, isPriceOf: nil
      def initialize(semanticId, date: nil, description: nil, name: nil, characteristicOf: nil, dimension: nil, value: nil, unit: nil, vatRate: nil, isPriceOf: nil)
        super(semanticId, date: date, description: description, name: name, characteristicOf: characteristicOf, dimension: dimension, value: value, unit: unit)
        @vat_rate = vatRate
        @is_price_of = isPriceOf
        self.semanticType = "dfc-b:Price"
        registerSemanticProperty("dfc-b:VATrate", &method("vat_rate")).valueSetter = method("vat_rate=")
        registerSemanticProperty("dfc-b:isPriceOf", &method("is_price_of")).valueSetter = method("is_price_of=")
      end

      # Original DFC v2 API aliases (see config/dfc-original-api.yaml).
      alias_method :vatRate, :vat_rate
      alias_method :vatRate=, :vat_rate=

      Core::SemanticObject.type_registry[SEMANTIC_TYPE] = self
    end
  end
end

File: typescript-connector/src/models/Price.ts
// Class from DFC Business Ontology: #Price
import { SemanticObject } from "../core/SemanticObject.js";
import { QuantitativeValue, type QuantitativeValueParams } from "./QuantitativeValue.js";

export interface PriceParams extends QuantitativeValueParams {
  vatRate?: number;
  isPriceOf?: string;
}
export class Price extends QuantitativeValue {
  static get SEMANTIC_TYPE(): string {
    return "dfc-b:Price";
  }

  vatRate?: number;
  isPriceOf?: string;

  constructor(
    semanticId: string,
    params?: PriceParams,
  ) {
    super(semanticId, params);
    this.vatRate = params?.vatRate;
    this.isPriceOf = params?.isPriceOf;
    this.semanticType = Price.SEMANTIC_TYPE;
    this.registerSemanticProperty("dfc-b:VATrate", () => this.vatRate);
    this.registerSemanticProperty("dfc-b:isPriceOf", () => this.isPriceOf);
  }
  static {
    SemanticObject.typeRegistry.set(Price.SEMANTIC_TYPE, Price);
  }
}
